### DISCUSSION TITLE
Boot on a device similar to raspberry pi?

### DISCUSSION THREAD

**Discussion by thorlucas** [phil-opp/blog_os#823] [General]
Hi,

Just started following along. Now unfortunately, my iMac refuses to boot the OS. I think it may have something to do with you saying UEFI was not supported? 

Additionally, I can't boot it on the raspberry pi 3 because that uses an ARM architecture, right?

Are there are any similar devices that would boot this OS? Or will it forever live in QEMU?

Thanks,

**Comment by thorlucas** [comment]
I should note that if anybody can find a way to boot my iMac or MacBook into it that would be very helpful. The startup manager does not recognize the USB as a bootable device unfortunately.

**Comment by phil-opp** [comment]
I'm working on UEFI support for the `bootloader` crate right now, so hopefully it will be bootable on UEFI machines soon. The ARM architecture is completely different to x86_64 and requires different code.

**Comment by thorlucas** [comment]
> The ARM architecture is completely different to x86_64 and requires different code.

That's what I expected!

You mentioned in your post that you can boot via USB. What devices have you had success with doing this?

Thanks for the great work, this is a very interesting exercise!

**Comment by phil-opp** [comment]
You're welcome, I'm glad to hear that you like the series.

Booting works on older machines that still support the traditional BIOS booting, before UEFI was a thing. I tried it on both an old desktop PC and laptop of mine and it worked without problems. My current laptop however only supports UEFI booting, so I can't run it on that machine either.

**Comment by YievCkim** [comment]
Hello,
 
@phil-opp do you plan to rewrite articles from the blog about what is changed for ARM. I think about the memory management that I expect different.

BTW, thank you for your serie of articles. I really learnt many thing, specifically on paging.

**Comment by thorlucas** [comment]
I think Phil said arm would require basically a total rewrite

**Comment by YievCkim** [comment]
Ok. I am not sure if it's a total rewrite of the code or a total rewrite of the code and articles in fact. I am curious how paging is implemented in ARM.

**Comment by Nicholas-Baron** [comment]
From my cursory glance, `rlibc` (which now points to the `compiler-builtins` crate) works on ARM. The `bootloader` and `x86_64` crates will need to add support for arm in the former case and removed entirely in the latter. I do not know about the other crates.

I have found other tutorials which cover Raspberry Pis specifically (some in C, at least one in Rust).

**Comment by ChocolateLoverRaj** [comment]
https://github.com/rust-embedded/rust-raspberrypi-OS-tutorials/ is useful. I think combining that tutorial with this tutorial, you could have some shared code and some specific code and make an OS that way.
